### PR TITLE
fix($resource): Delete $cancelRequest from toJSON()

### PR DESCRIPTION
**Pull request by hereblur**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Bug fix

**What is the current behavior? (You can also link to an open issue here)**
result of ngResource.toJSON still include $cancelRequest

**What is the new behavior (if this is a feature change)?**
remove delete $cancelRequest from ngResource.toJSON's result.

**Does this PR introduce a breaking change?**
No

**Please check if the PR fulfills these requirements**
- [ Y ] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [ N ] Tests for the changes have been added (for bug fixes / features)
- [ N ] Docs have been added / updated (for bug fixes / features)

**Other information**:

### CODE REVIEW

**Comment by gkalpak** [comment]
This for the PR. This was an intentional omission. Since `$cancelRequest` is always a function (and `JSON.stringify()` ignores functions), it shouldn't make any difference.

I am going to close this, but if you have a special usecase that breaks because of `$cancelRequest()`, please share below and I'll be happy to investigate.

**Comment by hereblur** [comment]
In my opinion, from the method name `toJSON()` we should be expected for a valid and ready to use of a  JSON object.

In many cases, people need the literal object from server without extra keys.
I got wrong number when I do `Object.keys( result ).length`.

I think one should not have to do `JSON.parse(JSON.stringifyr(result))` to get it works.

Thank you.

**Comment by gkalpak** [comment]
`toJSON` is a special method recognized by `JSON.stringify()` and is only intended to be used implicitly by that - i.e. not directly by the user ([MDN docs][1]).

That said, I guess it doesn't hurt to remove one more property ourselves (instead of relying on `JSON.stringify()`), so why not?

Reopening, but we need some tests for this (if we are making the change, we should make sure there will be no regressions in the future).

[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#toJSON()_behavior

**Comment by gkalpak** [comment]
Hm...I just realized `toJSON` is publicly documented and the docs state that the returned object does not contain any of the extra properties. So, we should definitely fix that. 

Thx @hereblur bringing it up.

**Comment by hereblur** [comment]
I signed it!

**Comment by hereblur** [comment]
@gkalpak Cool!, Thank you.
And I just pushed the test.

File: test/ngResource/resourceSpec.js [hunk 1]
```diff
@@ -750,6 +750,7 @@ describe('basic usage', function() {
     var json = JSON.parse(angular.toJson(cc));
     expect(json.$promise).not.toBeDefined();
     expect(json.$resolved).not.toBeDefined();
+    expect(json.$cancelRequest).not.toBeDefined();
```
Comment (gkalpak) [inline test/ngResource/resourceSpec.js]: Can you also add an expectation that is was there before? (Similar to what we have for `$promise`/`$resolved`.)

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: test/ngResource/resourceSpec.js (hunk 1 above)
Comment (gkalpak) [inline test/ngResource/resourceSpec.js]: I just realized that this test uses `JSON.parse(angular.toJson(cc))`, which means it would pass even without this PR.
Can you add a new test (similar to this one) that uses `cc.toJSON()` directly? (Leave this one as is - it doesn't hurt.)

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: test/ngResource/resourceSpec.js [hunk 2]
```diff
@@ -737,7 +740,7 @@ describe('basic usage', function() {
     expect(person2).toEqual(jasmine.any(Person));
   });
 
-  it('should not include $promise and $resolved when resource is toJson\'ed', function() {
+  it('should not include $promise, $resolved when resource is toJson\'ed', function() {
```
Comment (gkalpak) [inline test/ngResource/resourceSpec.js]: Revert this change.

File: test/ngResource/resourceSpec.js [hunk 3]
```diff
@@ -29,6 +29,9 @@ describe('basic usage', function() {
         }
       }
 
+    },
+    {
+      cancellable: true
```
Comment (gkalpak) [inline test/ngResource/resourceSpec.js]: Nit: It probably doesn't matter, but I would rather keep changes local, instead of affecting all tests. I.e. create a new, cancellable resource for your test.

**Review by gkalpak — CHANGES_REQUESTED** [review]

File: test/ngResource/resourceSpec.js [hunk 4]
```diff
@@ -753,6 +753,26 @@ describe('basic usage', function() {
...
+        method: 'GET',
+        cancellable: true
+      }
+    });
+
+    var creditCard = CreditCard.get();
+
+    expect(creditCard.$cancelRequest).toBeDefined();
+
+    $httpBackend.flush();
```
Comment (gkalpak) [inline test/ngResource/resourceSpec.js]: Is this necessary?

Comment (hereblur) [inline test/ngResource/resourceSpec.js]: Yes, we need to make sure It was there before remove it. same as the test next to it (for `$promise` and `$resolved`).
